### ISSUE TITLE
`drain` event trigger policy is wrong.

### ISSUE THREAD

**Issue by IvanProdaiko94** [jasonpincin/concurrent-queue#2]
I tried to use this lib, but faced the problem - there is no ability to find out when all of the tasks have been processed, or maybe I haven't found it yet.

Current behavior:
- Tasks pushed to the queue
- The concurrency limit reached
- Internal tasks buffer is flushed and all of the tasks from it trigger execution
- `drain` event triggered 'cause buffer is empty.
- wait till buffer will obtain enough task to launch the draining process.

Expected behavior:
- Tasks pushed to the queue.
- Start calling tasks from the start.
- Pop tasks one by one from the buffer when it is taken for processing.
- Do not flush the buffer completely.
- `drain` event triggered only in case buffer is empty.

Consider flowing data as a stream of tasks to queue. Using expected behavior you will get a notice when buffer will be empty, that is, turned out, to be called only once. That could be a more reliable solution for this pattern.

**Comment by jasonpincin** [comment]
Can you supply an example that exhibits this behavior? The expected behavior you outlined is exactly how this library is intended to work, and how it’s worked for me. 

It does not wait until concurrency is met to start processing. It will start processing as soon as something is placed on the queue. Everything is processed immediately until concurrency is reached. The `drain` event happens anytime the queue is emptied and all processing has been completed.

**Comment by IvanProdaiko94** [comment]
@jasonpincin, what I meant is:
Look [here](https://github.com/jasonpincin/concurrent-queue/blob/master/index.js#L77), you trigger flushing the queue, but, js is single threaded, and thus, you block the main thread and when you then write another task to queue the [drain](https://github.com/jasonpincin/concurrent-queue/blob/master/index.js#L75) event is called. Thus you end up with continuously calling `drain` event.

**Comment by jasonpincin** [comment]
The intention is that your processor function is not synchronous. This was intended to be used with an asynchronous, non-blocking professor function, otherwise you are correct, it will block. 

If you’d like to suggest an improvement over this, I’m happy to consider it.

**Comment by IvanProdaiko94** [comment]
I wrote data from the readable stream to queue and got this type of behavior(((
I will create a PR within a week
